### DISCUSSION TITLE
【PgFlow】改定リクエスト現象不明案件

### DISCUSSION THREAD

**Discussion by snext1220** [snext1220/stext#266] [Archive]
元々、https://github.com/snext1220/snext-management/issues/18#issuecomment-667667558 由来のIssueでしたが、備忘録的にdiscussionに移動しています。

**解決済みのもの、別Issueに移動済みのものを取り消し線でつぶしています（08/05）**

以下のものが現象再現せずでした。
当方の確認が悪い or ご意図を理解できていないのかもしれず、お手数ですが、詳細お教え戴いてもよろしいでしょうか（量が多くて申し訳ありません...）。
＃ スタッフの皆さま＞触られて、再現条件などお分かりになるものがあれば、是非お教えくださいm(_ _)m

> ~・各シーンの最初の空白（段落下げ）が反映されない？データをリロードするたびに段落下げがなくなっている。~

~［シナリオを実行］時にプレイ画面に反映されない、という意味ではないですよね...保存前後で空白が変化しないことまでは確認しております；先ほどUpload戴いたデータなどで、このシーンで、などあればお教えください。~

> ~・リンク追加を押して新しいシーン番号を指定しても、分岐が生成されない？~

~［シーン］タブの［リンク追加］ボタンでしょうか。
その場合、特定の条件でリンクが追加されない、ということでしょうか（常に？）。
現時点では問題なく動作しているように見えましたため、もし宜しければ、その時の操作詳細＆ブラウザーコンソールのキャプチャを戴いてもよろしいでしょうか（ご参考：https://sorcerian.hateblo.jp/entries/2018/06/18 ）~

> ・チャートでシーンを選択しても、シーンの表示がそのシーンになっていないことがある。知らずに編集してしまうとぐちゃぐちゃになる。

こちら今のところ再現できておりませんが、再現ケースなどございますでしょうか。

> ・気がつくと別のシーンのテキストが上書きされていることがある？？（再現性未検証）

とりあえずこちらでも弄ってまいりますが、上の件と絡んでいたりしますかね...？

> ・チャート上でシーンをドラッグしてフローを見る。それを繰り返すと、別のシーンをピックしてドラッグしてしまうことがある。

これもふたつ上の件と絡んでいるような気もしましたが、シーン同士の間隔を広げたい、などのご要望でしょうか。スミマセン、ご要望のご意図を図りかねたもので、お教え戴ければ幸いです。

> ・リンク追加して既存のシーンを指定しても、矢印が出ない。リンクのタブにはプルダウンメニューが表示されるが、選択できない。一度飛び先のシーンを削除してもう一度シーンを追加すると、矢印が生き返る。（シーンを削除しても、そのシーンとつながっているリンクの情報が消えていない？）

［シーン］タブの［リンク追加］ボタンからの操作でしょうか。
こちらでは、「20Sceneサンプル」テンプレートで1->12のように既存シーン同士でリンク追加しても特に問題なく追加されることまで確認しております。

全く急ぎではございませんので、お分かりのもの＆差し支えない範囲でお教え頂ければ幸いですm(_ _)m

**Comment by toki-sor1** [comment]
動作確認環境：　Win 8.1／84.0.4147.105(64bit)

> ・各シーンの最初の空白（段落下げ）が反映されない？データをリロードするたびに段落下げがなくなっている。

こちら、もしかしたら割と前からある不具合かもしれません……。
私はシナリオで段落下げを多用しているのですが、scenario.xmlを読み込むと、**各シーンの1行目のみ段落下げが無効になっていました。**

**Pg Flow上で作成・保存したscenario.xmlをPg Flowで読み込んだ後にも同上の状態になりました。**

ただ、当時は仕様と思い込んでいたのと、Pg Flowは主にフロチャ確認で使用していたため、不具合報告が遅れてしまいました；　申し訳ありません；；

> ・リンク追加を押して新しいシーン番号を指定しても、分岐が生成されない？

新しいシーンの番号についてですが、**新しいシーンを作成していない場合は、リンク追加時に新しいシーン番号を指定しても新しいシーンは生成されませんでした。**

例えば、リンク追加ボタンを押した後に「0->100（作成していないシーン番号）」と指定すると、「Scene 100へのリンク」は生成されるのですが、「Scene 100自体」は生成されないようでした。

＃「そういうことじゃないよ！」という回答になっていましたら申し訳ありません；

> ・チャートでシーンを選択しても、シーンの表示がそのシーンになっていないことがある。知らずに編集してしまうとぐちゃぐちゃになる。

>・気がつくと別のシーンのテキストが上書きされていることがある？？（再現性未検証）

私の環境では、上記2点の不具合は再現できていないですね。

> ・チャート上でシーンをドラッグしてフローを見る。それを繰り返すと、別のシーンをピックしてドラッグしてしまうことがある。

すみません、こちらは詳細を拝見してからのお返事になりそうです；

> ・リンク追加して既存のシーンを指定しても、矢印が出ない。リンクのタブにはプルダウンメニューが表示されるが、選択できない。一度飛び先のシーンを削除してもう一度シーンを追加すると、矢印が生き返る。（シーンを削除しても、そのシーンとつながっているリンクの情報が消えていない？）

私の環境では、こちらの不具合は再現できていないです～。

**Comment by snext1220** [comment]
@toki-sor1 さん：

お忙しい処、お教え戴きましてありがとうございます！

> Pg Flow上で作成・保存したscenario.xmlをPg Flowで読み込んだ後にも同上の状態になりました。

こちらですが手元でも確認しているのですが、やはり現象を確認できず、もう少しだけお教えください＞＜; 下のようなstext.json（Scene 0の冒頭に字下げの空白あり）を用意してみたのですが、問題なく空白が維持されているように見えます。
[stext.zip](https://github.com/snext1220/stext/files/5025820/stext.zip)
このようなことでは、ない、のですよね…
お手数ですが、問題の起こるファイルなどUp戴けると、大変助かります。。m(_ _)m

> 新しいシーンの番号についてですが、新しいシーンを作成していない場合は、リンク追加時に新しいシーン番号を指定しても新しいシーンは生成されませんでした。

こちらは私が思い切り勘違いしておりました！（シーン追加と混同しており、スミマセン；）
仰る通り、［リンク追加］は既存シーンへのリンク**だけ**の追加なので、そのようになりますね。。シーン作成を伴うリンク追加は［シーン追加］でお願いできますと。

ただ、この辺は挙動として判り難いのかなと思いますので、以下いずれかの対応を考えています。

1. ［リンク追加］では存在しないシーンを指定した場合、エラーを表示する 
2. ［リンク追加］で新規シーンの追加も可能にする

［リンク追加］は結構情報も多いダイアログになっているので、更に新規シーンのサマリなどまで入れさせるよりも、シンプルにエラーを出す方が良いかな（1. の方が良いかな）と思っておりますが、如何でしょうか。

**Comment by toki-sor1** [comment]
早速のお返事＆ご検証ありがとうございます！

> こちらですが手元でも確認しているのですが、やはり現象を確認できず、もう少しだけお教えください＞＜; 下のようなstext.json（Scene 0の冒頭に字下げの空白あり）を用意してみたのですが、問題なく空白が維持されているように見えます。

ご用意いただきましたstext.zip内にあるstext.jsonで検証してみました。

- 読込：stext.json→保存・変換：stext.jsonでは、一行目の段落下げが有効になっていた
- 読込：stext.json→保存・変換：**scenario.xml**では、一行目の段落下げが**無効**になっていた

どうやらjson形式ファイルのまま保存すれば段落下げが保たれているのですが、**xml形式ファイルに変換して保存したファイルの場合は、読み込むと一行目の段落下げが無効になるようでした。**
こちら（[stext_error.zip](https://github.com/snext1220/stext/files/5027957/stext_error.zip)）が変換後（不具合が再現できた）ファイル＋暗き沼scenario.xmlになります。
＃そのままテキストエディターで開いていただくと、一行目の段落が下がっているのを確認できます

> 仰る通り、［リンク追加］は既存シーンへのリンクだけの追加なので、そのようになりますね。。シーン作成を伴うリンク追加は［シーン追加］でお願いできますと。

仕様のご解説ありがとうございます！
なるほど、「編集するシーン」の一覧に新しいシーンが追加されていなかった理由は、そういうことなのですね～。
（勘違いで見当違いのことを書いていないか不安だったので、安心しました…ｗ）

対応については、**1. ［リンク追加］では存在しないシーンを指定した場合、エラーを表示する**が良いなと思っています。
理由は、既存のシーン番号を指定しようとした時に、誤って新しいシーン番号を入れてしまう可能性があるかも、と思ったからです～（意図しない新規シーン追加＆リンクミス）。

**Comment by snext1220** [comment]
@toki-sor1 さん：

ご確認ありがとうございます！助かりました。
.xmlファイルの出力までは見ていたのですが、.xmlファイルから.jsonファイル（PgFlow）への変換で問題が起きていたとは盲点でした；（いいわけですが、.jsonがマスターという感覚でいたもので、PgFlowで作成したファイルを.xml->.jsonすることがなかったのが穴でした...）。 #217 にてBug Issueも立てて、対応済みでございます～

［リンク追加］の件もありがとうございます！（1. 個人的には賛成です～）
こちらは #218 に移動し、他の皆さまのご意見を伺ったうえで方針確定してまいりますね！

**Comment by toki-sor1** [comment]
動作確認環境：　Win 8.1／84.0.4147.105(64bit)

今度は.xmlファイルの一行目の段落が意図したとおり下がっていることを確認しました！
早速のご対応ありがとうございました＆お疲れ様でした！

> .jsonがマスターという感覚でいたもので、PgFlowで作成したファイルを.xml->.jsonすることがなかったのが穴でした...

私の場合は、.jsonファイルはテキストエディターで上手く編集できないので、いつも.xmlファイルに変換して保存していたのですが、なるほど、これは話が噛み合わない訳ですねｗ
#217 立てもありがとうございました！　こちらの返信はこの後に！

> ［リンク追加］の件もありがとうございます！（1. 個人的には賛成です～）
こちらは #218 に移動し、他の皆さまのご意見を伺ったうえで方針確定してまいりますね！

こちらもありがとうございました！　方針確定の件も了解です～！

**Comment by snext1220** [comment]
こちらこそ早速のご確認ありがとうございました！
最初期からの問題だったと思うので、ここで解決してよかったです～
＃ #217 のお返事もありがとうございます！

> 私の場合は、.jsonファイルはテキストエディターで上手く編集できないので、いつも.xmlファイルに変換して保存していたのですが、～

成程、.xmlファイルでの並行編集が結構発生している感じですかね。
この辺、もし宜しければ、どの辺りを生xml編集しているかお教え戴いてもよろしいでしょうか（勿論、全然急ぎではございませんです～）。この辺、思い切りPgFlowの改良アイデアの宝庫のような気がしたものでｗ

**Comment by olreia** [comment]
検証いただき誠に恐れ入ります。
上に挙げた不具合の中で再現できていないものについては、とりあえず起きた現象をそのままメモしたのですが、私の勘違いか操作ミス等の可能性も十分にあります。申し訳ありません。

・チャートでシーンを選択しても、シーンの表示がそのシーンになっていないことがある。知らずに編集してしまうとぐちゃぐちゃになる。
・気がつくと別のシーンのテキストが上書きされていることがある？？（再現性未検証）
これらは本当に「気がつくと」なので、情けない限りなのですが、どうもフローチャートとシーンタブを交互に弄っているときに、チャートの特定のシーンをクリックしてもシーンタブがそのシーンのものに遷移していないことがあるのではないかと思っています。私の環境の問題かもしれませんが。。

・チャート上でシーンをドラッグしてフローを見る。それを繰り返すと、別のシーンをピックしてドラッグしてしまうことがある。
これは本当に些末な話で、シーン同士の繋がりを確認するためにシーンをドラッグして左右に振って矢印を揺らす作業を繰り返しているときに、いつの間にか別のシーンをドラッグしていることがあったというものです。単にマウスボタンから一瞬手が離れただけかも……。

・リンク追加して既存のシーンを指定しても、矢印が出ない。リンクのタブにはプルダウンメニューが表示されるが、選択できない。一度飛び先のシーンを削除してもう一度シーンを追加すると、矢印が生き返る。（シーンを削除しても、そのシーンとつながっているリンクの情報が消えていない？）
すみません、こちらは今試してみたところ再現できませんでした。
起こったことをありのままに書いたように記憶しているのですが……。上のものと同じで、チャートとタブの同期にズレがある瞬間があったのかもしれません。

とりあえずこの辺で。
このたびは大変お世話になりました。
また次回作で使い込んでみたいと思います。

**Comment by snext1220** [comment]
お忙しい処、ご確認ありがとうございます！

現在再現しないものについて、当方でも現時点では確認できず申し訳ありません。ただ
、なにかしら発生条件があるものもありそうなので、**備忘録用に本Issueは残し、今後の皆さまからの情報提供を待ちたい**と思います（もちろん、私も気にしてみてまいります）。恐縮ですが、関連しそうな現象あった場合には本Issueでお教え戴けますと幸いです＞皆さま

現時点では曖昧な回答になってしまい恐縮ですが、引き続き要件確定したIssueから順に対応を進めていければと思いますので、暫しお待ちくださいませ。
